src/main/java/org/snomed/snowstorm/rest/ConceptController.java

```diff
@@ -580,6 +589,143 @@ public ExpressionStringPojo getConceptNormalForm(
 
 		return new ExpressionStringPojo(expression.toString(includeTerms));
 	}
+	
+	@PostMapping(value = "/{branch}/concepts/copy")
+	@PreAuthorize("hasPermission('AUTHOR', #branch)")
+	@JsonView(value = View.Component.class)
+	public Collection<ConceptMini> copyConcepts(
+			@ApiParam(value = "Destination branch where the concepts are copied to.")
+			@PathVariable String branch,
+			@ApiParam(value = "Source branch where the concepts are selected from.")
+			@RequestParam String sourceBranch,
+			@ApiParam(value = "ECL expression for selecting concepts to copy between the branches.")
+			@RequestParam String ecl,
+			@ApiParam(value = "Include dependant components. Defaults to false.")
+			@RequestParam(required = false, defaultValue = "false") boolean includeDependencies) throws ServiceException {
+		String sourceBranchPath = BranchPathUriUtil.decodePath(sourceBranch);
+		String destinationBranchPath = BranchPathUriUtil.decodePath(branch);
+
+		final List<LanguageDialect> languageDialects = List.of(new LanguageDialect(Config.DEFAULT_LANGUAGE_CODE, Long.parseLong(Concepts.US_EN_LANG_REFSET)));
+
+		// Get concepts
+		QueryService.ConceptQueryBuilder queryBuilder = queryService.createQueryBuilder(true)
+				.activeFilter(true)
+				.ecl(ecl)
+				.isReleased(true)
+				.resultLanguageDialects(languageDialects);
+
+		List<ConceptMini> conceptsFound = queryService.search(queryBuilder, sourceBranchPath, LARGE_PAGE).getContent();
+		List<ConceptMini> conceptsCreated = new ArrayList<>();
+
+		for (ConceptMini concept : conceptsFound) {
+			ConceptMini conceptCreated = copyConcept(concept, sourceBranchPath, destinationBranchPath, languageDialects, includeDependencies);
+			conceptsCreated.add(conceptCreated);
+		}
+		return conceptsCreated;
+	}
+
+	private String getModuleId(String branchPath) {
+		Branch branch = branchService.findBranchOrThrow(branchPath, true);
+		final boolean isExtension = (branch.getMetadata() != null && !Strings.isEmpty(branch.getMetadata().getString(BranchMetadataKeys.DEPENDENCY_PACKAGE)));
+
+		return isExtension ? branch.getMetadata().getString(BranchMetadataKeys.DEFAULT_MODULE_ID) : Concepts.CORE_MODULE;
+	}
+
+	private ConceptMini copyConcept(ConceptMini concept, String sourceBranchPath, String destinationBranchPath, List<LanguageDialect> languageDialects, boolean includeDependencies) throws ServiceException {
```

Comment:
The method name is copyConcept but the actual implementation is more than just copy a concept from one branch to another. Maybe donateConcept or transferConcept might be more suitable.

Reply:
There is tones of business logic here. This method should be moved to a service rather than living in a controller.

Reply:
@kaicode I tried that, but I got an issue with context not found :-( I think it did not like me adding referenceSetMemberService to the ConceptService class, but I am not sure.

Reply:
@kaicode I think that @Lazy did the trick. It appears to be working now. Has moved `donateConcept()` (aka `copyConcept()`) and `getModuleId()` into `ConceptService` class.

Reply:
@astro-snail sorry I missed these comments. I'm glad Lazy works. A separate "concept donation service" is an alternative to get around the wiring issue but sounds like you don't need it.